[cxxmodules][cling] Avoid loading some unnecessary modules

smuzaffar opened the pull request:
testign https://github.com/root-project/root/pull/10910 for cmssw

guitargeek committed:
Avoid using deperecated `RooCatType` in RooStats tutorials

With the rewrite of the RooCategory classes, the `RooCatType` was
deprecated (moved into RooFitLegacy) and should also not be used anymore
in tutorial code.

lmoneta committed:
[RF] Evaluate pdf in RooRealIntegral only if the normalization set is defined

vgvassilev committed:
Simplify CallFunc. NFC

This patch reduces the code bloat in TClingCallFunc trying to make the code
more readable.

vgvassilev committed:
Use simplisticCastAs

vgvassilev committed:
Remove duplicate sv_to as we have cling::Value::simplisticCastAs

linev committed:
Correctly use Copy method in TF[1,2,3] functions classes

One should explicitly specify which method is used

linev committed:
Call proper Copy method with TH[1,2,3] classes

linev committed:
Use proper Copy method in TProfile[,2D,3D] classes

linev committed:
Mark copy constructors for TH1, TH2, TH3 classes as deleted

They should not be used at all

linev committed:
Improve syntax in TH1.h and TH2.h

Use nullptr, remove dummy ";"

linev committed:
Move implementation of static methods from TH3.h

It is always dangerous because of potential linking problems

vgvassilev committed:
We don't need advanced normalization for building a call wrapper.

This patch also adds shared_ptr callfunc test.

GetNormalizedName was introduced in root-project/root@291f23e to replace
root-project/root@a211a41386e46f8467fdd0c2230d4432d521f542 which taught clang
to drop the `shared_ptr` default argument as it did not and thus callfunc failed
to compile its wrapper.

This test ensures we do not regress. It produces a callfunc wrapper of the type:
__attribute__((used)) __attribute__((annotate("__cling__ptrcheck(off)")))
extern "C" void __cf_11(void* obj, int nargs, void** args, void* ret)
{
   ((void (&)(std::shared_ptr<std::vector<E, std::allocator<E> > >))add_sp)(*(std::shared_ptr<std::vector<E, std::allocator<E> > >*)args[0]);
   return;
}

guitargeek committed:
[RF] Clean duplicate include in `stressRooFit_tests.h`

sanjibansg committed:
feat: support for parsing binary operators in Keras parser

guitargeek committed:
[RF] Don't forget to call `redirectServersHook()` of the base class

If one overloads `RooAbsArg::redirectServersHook()`, one also needs to
call the `redirectServersHook()` of the base class. Otherwise, important
things that must be done when redirecting the servers of a `RooAbsReal`
or `RooAbsPdf` for example are not done.

In particular, this fixes some ASAN build test failures where the
`RooAbsPdf::_normSet` member was not reset when redirecting servers
because `RooAbsPdf::redirectServersHook()` was not called.

hahnjo committed:
Delete nop method TMVA::SVWorkingSet::PrintStat()

It was doing nothing, so I expect no user to have ever called it.

hahnjo committed:
Remove more unused variables in various areas

All found by Clang's -Wunused-but-set-variable.

eguiraud committed:
[DF] Apply some include fixes from include-what-you-use

guitargeek committed:
[RF] Avoid ASAN build failures in two RooFitCore unit tests

In the ASAN build, the `testNaNPacker` and `testRooBinSamplingPdf` tests
failed because the normalization sets of the PDFs were destructed too
early. This commit is adapting these tests such that it doesn't happen
anymore.

eguiraud committed:
[NFC][DF] Add note about `return` statements in string expressions

eguiraud committed:
[NFC][DF] Properly link to Define method from docs

eguiraud committed:
[NFC][DF] Fix typo in docs

osschar committed:
[GL] Add scaling factors for width and bounding-box size used when drawing simple axes in TGLViewer.

guitargeek committed:
[RF] Use less manual memory allocation in RooAbsReal

guitargeek committed:
[RF] Fix ASAN build failure related to norm sets in RooIntegralMorph

In the `RooIntegralMorph` cache, the normalization set to evaluate the
PDF was created on the stack, and then the proxies in this PDF pointed
to this `RooArgSet` that gets destructed at the end of the function.
Now, the normSet pointers in the proxies are invalid, which caused ASAN
build failures.

This can be easily fixed by using the `RooIntegralMorph::_nset` member
instead, which has the same variables as the temporary norm set had.

guitargeek committed:
[RF] Implement `RooRealIntegral::fillNormSetForServer`

In pass-through mode, the RooRealIntegral evaluates the PDF with a
fixed normalization set that also needs to be reported to the BatchMode
via the `fillNormSetForServer` interface.

osschar committed:
Move implementation of static member functions to cxx file.

couet committed:
[skip=-ci] \ was not displayed in ref guide. (#11184)

lmoneta committed:
[tmva] Fix tutorials for failures seen in CI (#11185)

We need to veto RegressionKeras.py since it is to oslow in evaluating the model, since it is doing on an event by event base

guitargeek committed:
[RF] Reset cached normalization sets when redirecting servers

When servers are redirected, the cached normalization sets in the RooFit
objects are most likely invalid. It also doesn't hut to reevaluate them,
because evaluation happens much more frequent than server redirection.

As far as I know, there are three places where normalization sets are
cached:

  1. The `RooAbsPdf::_normSet` member

  2. The `RooAbsReal::_lastNSet` member

  3. The `RooAbsProxy::_nset` member

The `RooAbsPdf::_normSet` is already reset to `nullptr` in the
`redirectServersHook` of RooAbsPdf.

This commit completes that behavior by also resetting the other members
in the `redirectServersHook` of their respective class.

This change fixes some failures in the ASAN build.

vgvassilev committed:
Use the information in cling::Value, drop fReturnIsRecordType

Neel-Shah-29 committed:
[GSOC][TMVA][SOFIE] Cast ONNX Operator implemented with the corresponding unit tests (#11033)

* Cast ONNX Operator implemented with the corresponding unit tests

Added the functionality and support of int input type in Cast ONNX Operator

* Attribute type added to ROperator Cast Class and modified the RModel Parser for supporting different input types

* The functionality and support for other datatypes added for the cast operator and added the support to RModel::Generate method also

* made the required changes related to support for other datatypes

* Extended the support for other datatypes in the infer function

* Required changes made related to support of different datatypes in SOFIE

* Apply various fixes to support different input/output type. This fixes the new Cast operator.
   Several changes are needed for Cast since the input tensor can be of a type different than float
   Apply also a fix for parsing correctly the attribute of Cast

* The attribute fattr_type changed to fAttrType

Co-authored-by: moneta <lorenzo.moneta@cern.ch>

guitargeek committed:
[RF] Fix RooSimultaneous parameter redirection in BatchModeHelpers

In the BatchModeHelpers that are used to create the NLL object with the
BatchMode, the PDF is cloned if it is a RooSimultaneous.

After cloning, the cloned parameters are reattached to the original
model. However, it should be the other way around because otherwise the
original model points to invalid parameter args when the fit is done.

guitargeek committed:
[RF] Use less manual memory management in BatchModeHelpers

This is not a fix to any issue, just a refactoring to avoid manual
memory management in the BatchModeHelpers.

guitargeek committed:
[RF] Fix invalid memory access in RooFitDriver destructor

In the `RooFitDriver::NodeInfo` destructor, the data tokens for all
nodes are reset. However, if this happens after the destruction of the
normalization integrals in the computation graph, it tries to reset the
data token for nodes that don't exist anymore.

Hence, the data token resetting is better done at the beginning of the
RooFitDriver destructor.

HDembinski committed:
[Math][Minuit2] IsAboveMaxEdm must return true if Edm is NaN (#11159)

guitargeek committed:
[RF] Use `recursiveRedirectServers()` in NormalizationHelpers

In the `foldIntegrals()` function that reverses the unfolting of
normalization integrals in the computation graph, some custom logic was
used to redirect the servers. However, it caused some failures in the
ASAN builds because it accessed invalid memory.

This commit suggests to fix that by using the already existing
`RooAbsArg::recursiveRedirectServers()` function for the same logic.

Furthermore, in `unfoldIntegrals()`, clients outside the computation
graph are not considered for the server redirection, which also fixes
ASAN build failures.

guitargeek committed:
[RF] New mechanism to detect if _normSet in RooAbsPdf is invalid

In RooAbsPdf, the following check is done to check if the normalization
set has changed:
```C++
RooFit::getUniqueId(nset) != RooFit::getUniqueId(_normSet)
```

This works, but like the previous pointer comparison it still gives
false results if the `_normSet` was replaced buy another RooArgSet at
the same memory location. The problem can be reproduced with this code:

```C++
using namespace RooFit;

// Create observables
RooRealVar x("x", "x", -5, 5);
RooRealVar y("y", "y", -5, 5);

// Create signal pdf gauss(x)*gauss(y)
RooGaussian gx("gx", "gx", x, RooConst(0), RooConst(1));
RooGaussian gy("gy", "gy", y, RooConst(0), RooConst(1));
RooProdPdf sig("sig", "sig", {gx, gy});

// Create composite pdf
RooAddPdf model("model", "model", RooArgList(sig), RooConst(100.));

// The results of the last two lines depend on the order of execution
// which should not be the case!
std::cout << model.getVal(x) << std::endl;
std::cout << model.getVal({x, y}) << std::endl;
```

The solution would be to replace the `_normSet` member directly with
its unique ID value, but then it could not be used anymore. Hence we
need both.

This commit also adds a new private `RooAbsPdf` member functions to set
the `_normSet` together with the new `_normSetId` member, and one
protected `isActiveNormSet(RooArgSet const*)` function. The latter
function is comparing the input to the current `_normSet`, but shortcuts
to `false` is the ID does not match with the last `_normSet` ID. Like
this, we avoid dereferencing invalid `_normSet` pointers.

This change is mainly done to fix several of the ASAN build failures in
RooFit.

vgvassilev committed:
fIgnoreExtraArgs is set but not used, remove it.

vgvassilev committed:
Reduce duplicated code in the tests; add missing dependency.

cmsbuild commented:
A new Pull Request was created by  @smuzaffar (Malik Shahzad Muzaffar) for branch cms/master/6018b5fd45.

@cmsbuild, @smuzaffar, @aandvalenzuela, @iarspider can you please review it and eventually sign? Thanks.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.
cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smuzaffar commented:
please test for CMSSW_12_5_ROOT6_X

HannaOlvhammar committed:
[RF] Enable vectorized first order interpolation for weight evaluation

With the function `RooDataHist::interpolateLinear` it is now possible to use
`RooDataHist::weights()` (implemented in #11171) to evaluate weights for one
dimensional histograms with up to first order interpolation.

`RooDataHist::interpolateLinear` finds the coefficients of a straight line
between two neighboring bin centers by solving a system of two linear
equations. With the coefficients of the straight line, the interpolated weight
for an event between two bin centers can be calculated.

If an event in a bin is to the left of the bin center coordinates, the
corresponding weight is found by interpolating between the current bin and the
neighboring bin to the left. If the event is to the right of the bin center,
the weight is found by inteperpolating between the current bin center and the
one to the right. For the first and last bin, the interpolation is performed
either by mirroring the histogram outside its boundaries, or by applying cdf
boundaries.

`RooDataHist::weights()` has been updated so that it calls
`RooDataHist::interpolateLinear()` when the interpolation order is one.
Furthermore, `RooDataHist::computeBatch()` and `RooHistFunc::computeBatch()`
have been updated so that they call `RooDatahist::weights()` in the cases
where the histogram is one dimensional and the interpolation order is zero or
one. Unit tests have been added for both zeroth and first order interpolation.

guitargeek committed:
[RF] Clear normalization cache in `RooAbsPdf::redirectServersHook`

If servers are redirected, the cached normalization integrals and
normalization sets are most likely invalid.

This change fixes failures in the ASAN build tests.

guitargeek committed:
[RF] Always adjust proxies when normalization set instance has changed

In the past, the proxies where not adjusted when the cached integral
object didn't change.

However, the cache checks if the nset was already cached **by content**,
and not by RooArgSet instance! So it can happen that the normalization
set object is different, but the integral object is the same, in which
case it would be wrong to not adjust the proxies. They always have to be
adjusted when the nset changed, which is always the case when
`syncNormalization()` is called.

The `nsetChanged` varaibles were also renamed to `nintChanged` changed,
because it tells if the integral object changed, not the the
normalization set object.

linev committed:
Intentially call virtual Copy method in TH[1,2,3]D classes

These classes are base for TProfile[,2D,3D] and copy constructor
should not be used together with such classes. This is check
if TProfile::Copy() method.

hahnjo committed:
[tmva][sofie] Add missing cassert includes (#11201)

ikabadzhov committed:
[DF] Use strings instead of html in SaveGraph labels

guitargeek committed:
[RF] Make `RooCmdConfig::stripCmdList` a static function

There was no reason why it should not be a static function to begin
with, and it will also not cause any backwards incompatibilites because
a static function can also be called from an instance.

guitargeek committed:
[RF] Remove `FillColor` command for single curves in error band plots

When plotting the single curves in `RooAbsReal::plotOnWithErrorBand()`,
the `"FillColor"` command should be removed because it has no effect on
plotting single curves and would cause warnings, as can be seen when
running the `rf610_visualerror` tutorial.

Closes #9859.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16db43/26875/summary.html
**COMMIT**: 951518cb78ea9f2adff9e3d8a663fabe613c8054
**CMSSW**: CMSSW_12_5_ROOT6_X_2022-08-16-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/root/173/26875/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testTauEmbeddingProducers had ERRORS
</pre>

## Comparison Summary



There are some workflows for which there are errors in the baseline:
 136.9 step 4
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 16478 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3692500
 - DQMHistoTests: Total failures: 37935
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3654542
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 2.372 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 11834.0 ): 2.372 KiB        SiStrip/MechanicalView
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-16-2300+16db43/52182/triggerResults): found differences in 5 / 50 workflows

hahnjo committed:
[eve7] Add missing includes to REveElement.hxx

hahnjo committed:
[eve7] Include json_fwd.hpp if available

Version 3.11 of nlohmann/json introduced "versioned, ABI-tagged inline
namespace"s, which breaks our forward declaration. Fortunately, we can
assume the json_fwd.hpp header to be present starting from that same
version because the JSON_MultipleHeaders option now defaults to ON and
even if not, json_fwd.hpp is installed since patch version 3.11.2. For
earlier versions, both methods work but json_fwd.hpp isn't guaranteed
to be installed. Still use it if available.

Fixes #11130

hahnjo committed:
[cmake] Check for json_fwd.hpp in versions > 3.11

Our forward declaration in REveElement.hxx breaks with the versioned
namespaces in 3.11, so we require the json_fwd.hpp header.

linev committed:
[webgui] let enable standard browser usage with web display

If --web=standard or --web=std is configured, default web
browser will be started. Can only be used with interactive
(non-batch) sessions. Also --web=browser can be used

linev committed:
[webgui] let configure web display with WebGui.Display parameter

Such parameter can be add to .rootrc file

linev committed:
[webgui] let configure web display via ROOT_WEBDISPLAY env var

It is evaluated as very first

linev committed:
[webgui] Update/improve documentation

linev committed:
[webgui] use default instead of standard as name for dflt browser

lmoneta committed:
[math][fit] Fix Fitter class for temporary FitData (issue #11154)

The Fitter::Fit interfaces can accept a reference to a FitData object or a
shared_ptr to the FitData. In the first case if the give data object was deleted by the user,
the Fitter was containing a pointer ot a deleted object.
This is fixed now bby copying the input FitData in the Fitter class when a reference is provided by the user.
If one wants to avoid the copying, the interface passing the `shared_ptr` should be used.

This PR fixeds also:

 - copying of BinData and UnBinData classes. The UnBinData class was missing a public copy constructor/assignment,
   and the copy of BinData/FitData was missing to copy a data member.
 - copying of FitResult. The FitResult does not need to implement a copy constructor, it can use the default one.
   Before the fFitData member was not copied.
 - Improves doxygen documentation

lmoneta committed:
[math][fit] Add correct copy constructor/assignment for SparseData class

Use also a unique_ptr instead of a raw pointer

linev committed:
[jsroot] update to 7.2.0 with few extra fixes

lmoneta committed:
[math][fit] Fix a buf in the copy constructor/assignment of BinData

Apply also the review comments by Jonas.

Remove also teh DummyDeleter used to construct shared_ptr from external
references of FitData classes
It has ben decided that when external referece are passed to the Fitter or the
FCN classes, the data are copied in. If one wants to avoid copying it should pass a shared_ptr instead of a simple reference.

linev committed:
[webgui] introduce WebGui.OnetimeKey rootrc parameter

If specified, enforce unique key every time client connect to
web window. Once key used, it not possible to reconnect.
If one wants to connect to window again, one should
call RWebWindow::Show() method again.

linev committed:
[webgui] add warning message when web browser started on remote

Checking DISPLAY variable, one could identify if X11 forwarding
is configured. Show warning that ssh tunnel and --web=server mode
can be used instead.

bellenot committed:
Fix potential crash with RBrowser on Windows

Fix potential crash when quitting ROOT from RBrowser on Windows and simplify the code used to close the TBrowser main Window

HannaOlvhammar committed:
[RF] Enable vectorized second order interpolation for weight evaluation

With the function `RooDataHist::interpolateQuadratic()` it is now possible to
use `RooDataHist::weights()` (implemented in #11171) to evaluate weights for
one dimensional histograms with up to second order interpolation.

The structure of `RooDataHist::interpolateQuadratic()` is similar to that of
`RooDataHist::interpolateLinear()`, which was implemented in #11211
It finds the coefficients of a second order polynomial between three
neighboring bin centers by solving a system of three quadratic equations. With
the coefficients of the quadratic function, the interpolated weight for an event
between two bin centers can be calculated.

If an event in a bin is to the left of the bin center coordinates, the
corresponding weight is found by interpolating between the current bin and the
two neighboring bins to the left. If the event is to the right of the bin center,
the weight is found by interpolating between the current bin center, one bin to
the the left and one bin to the right. For the first two bins and the last bin,
the interpolation is performed either by mirroring the histogram outside its
boundaries or by applying cdf boundaries.

`RooDataHist::weights()` has been updated so that it calls
`RooDataHist::interpolateQuadratic()` when the interpolation order is two.
Furthermore, `RooDataHist::computeBatch()` and `RooHistFunc::computeBatch()`
have been updated so that they always call RooDatahist::weights() in the case
where the histogram is one dimensional. Since the old interpolation
implementation `RooDataHist::interpolateDim()` only supported up to second
order interpolation, it has been completely replaced by
`RooDataHist::interpolateLinear()` and `RooDataHist::interpolateQuadratic()`
in these `computeBatch()` functions in the case of one dimensional histograms.

eguiraud committed:
[IO] Avoid nullptr dereference in I/O of std::pair

cmsbuild commented:
Pull request #173 was updated.

guitargeek committed:
Fix the classic double "the the" typos in the ROOT documentation

hahnjo committed:
[asan] Only preload ASan runtime for CMD (#11226)

Some Python tests define PRE_ and POST_CMD; they should not get the
ASan runtime preloaded. Invent a new CMD_ENV that supports exactly
one environment variable that is applied to CMD. This fixes another
16 test failures.

bellenot committed:
[cmake] Add POLICY CMP0135 and set it to NEW

Fixes the following warning with CMake 3.24:
```
CMake Warning (dev) at /usr/share/cmake/Modules/ExternalProject.cmake:3071 (message):
  The DOWNLOAD_EXTRACT_TIMESTAMP option was not given and policy CMP0135 is
  not set.  The policy's OLD behavior will be used.  When using a URL
  download, the timestamps of extracted files should preferably be that of
  the time of extraction, otherwise code that depends on the extracted
  contents might not be rebuilt if the URL changes.  The OLD behavior
  preserves the timestamps from the archive instead, but this is usually not
  what you want.  Update your project to the NEW behavior or specify the
  DOWNLOAD_EXTRACT_TIMESTAMP option with a value of true to avoid this
  robustness issue.
```

Neel-Shah-29 committed:
[GSOC][TMVA][SOFIE] Pow ONNX Operator added to SOFIE with the corresponding unit tests. (#10971)

* Pow ONNX Operator added with the corresponding unit tests to validate the code.

* Pow ONNX Operator implemented as one of the Binary Operator enum to reduce the code

* Fixed the merge conflicts related to Pow Operator

cmsbuild commented:
Pull request #173 was updated.

guitargeek committed:
[RF] Refactor CUDA helpers in RooFitDriver to free functions

This is done to better understand when a function changes the state of
the RooFitDriver.

guitargeek committed:
[RF] Ignore empty spans when filling server values in computeBatch

In `RooAbsReal::computeBatch()`, all server values are filled from the
dataMap. However, it can happen that some values are not in the dataMap
if the server is not a value or shape server. These servers should be
skipped.

guitargeek committed:
[RF] Do initial pass in RooFitDriver::getValHeterogeneous() on GPU

So far, the first pass to measure the timing was done on the CPU and the
second on the GPU. This is not practical when only doing one evaluation
for development purposes. Hence the order is now reversed.

cmsbuild commented:
Pull request #173 was updated.

cmsbuild commented:
Pull request #173 was updated.

cmsbuild commented:
Pull request #173 was updated.

eguiraud committed:
[NFC] Make nbviewer/SWAN badges more prominent in the doxygen docs

couet committed:
[skip-ci] improve error message (#11235)

bellenot committed:
Disable brotli support feature in Freetype (not only on Apple) (#11242)

* Disable brotli support feature in Freetype (not only on Apple)

Should fix #11239

* Remove code instead of commenting it out

junaire committed:
[cxxmodules][cling] Avoid loading some unnecessary modules

Signed-off-by: Jun Zhang <jun@junz.org>

cmsbuild commented:
Pull request #173 was updated.
